petl/io/gsheet.py

```diff
@@ -0,0 +1,156 @@
+            except gspread.exceptions.SpreadsheetNotFound:
+                wb = gspread_client.open(self.filename)
+
+        # Allow for user to specify no sheet, sheet index or sheet name
+        if self.worksheet_title is None:
+            ws = wb.sheet1
+        elif isinstance(self.worksheet_title, int):
+            ws = wb.get_worksheet(self.worksheet_title)
+        else:
+            # use text_type for cross version compatibility
+            ws = wb.worksheet(text_type(self.worksheet_title))
+
+        # grab the range or grab the whole sheet
+        if self.range_string:
+            # start_cell -> top left, end_cell -> bottom right
+            start_cell, end_cell = self.range_string.split(':')
+            start_row, start_col = gspread.utils.a1_to_rowcol(start_cell)
+            end_row, end_col = gspread.utils.a1_to_rowcol(end_cell)
+            # gspread starts its indices at 1
+            for i, row in enumerate(ws.get_all_values(), start=1):
+                if i in range(start_row, end_row + 1):
+                    start_col_index = start_col - 1
+                    yield tuple(row[start_col_index:end_col])
+        else:
+            # no range specified, so return all the rows
+            for row in ws.get_all_values():
+                yield tuple(row)
+
+
+def togsheet(tbl, filename, credentials, worksheet_title=None,
+             share_emails=[], role='writer'):
+    """
+    Write a table to a new google sheet.
+
+    `filename` will be the title of the workbook when uploaded to google sheets.
+
+    `credentials` are used to authenticate with the google apis.
+    For more info, visit: http://gspread.readthedocs.io/en/latest/oauth2.html
+
+    If `worksheet_title` is specified, the first worksheet in the spreadsheet
+    will be renamed to the value of `worksheet_title`.
+
+    The spreadsheet will be shared with all emails in `share_emails` with
+    `role` permissions granted.
+    For more info, visit: https://developers.google.com/drive/v3/web/manage-sharing
+
+    Note: necessary scope for using togsheet is:
+         'https://spreadsheets.google.com/feeds'
+         'https://www.googleapis.com/auth/drive'
+
+    Example usage::
+        >>> import petl as etl
+        >>> from oauth2client.service_account import ServiceAccountCredentials
+        >>> scope = ['https://spreadsheets.google.com/feeds',
+                     'https://www.googleapis.com/auth/drive']
+        >>> credentials = ServiceAccountCredentials.from_json_keyfile_name('path/to/credentials.json', scope)
+        >>> cols = [[0, 1, 2], ['a', 'b', 'c']]
+        >>> tbl = etl.fromcolumns(cols)
+        >>> etl.togsheet(tbl, 'example', credentials)
+    """
+
+    import gspread
+    gspread_client = gspread.authorize(credentials)
+    spreadsheet = gspread_client.create(filename)
+    rows = len(tbl)
+    # get the max length and add [0] to take care of empty iterables
+    cols = max([0] + [len(row) for row in tbl])
```

Comment:
So this probably doesn't matter much because most users working with gsheets probably won't be working with large tables, but there are two full passes here over the input table and these could be combined into a single pass. I.e., you could iterate once over the input table to obtain both the number of rows and the number of columns. I.e., something like:

```
nrows = 0
ncols = 0
for row in tbl:
    nrows += 1
    ncols = max(ncols, len(row))
```